### DISCUSSION TITLE
Episode 164: Are We Going to Run Out of Credit Cards?

### DISCUSSION THREAD

**Discussion by codereport** [codereport/adsp2#56] [Episode]
Episode Link: https://adspthepodcast.com/2024/01/12/Episode-164.html

Feel free to write comments or ask questions 🙂

**Comment by pascal754** [comment]
This episode was hilarious.

There is a problem to check whether a credit card number is valid using the Luhn algorithm in CS50:

https://cs50.harvard.edu/x/2024/psets/1/credit/

**Comment by VincentZalzal** [comment]
To answer the question, it definitely wasn't the best episode, but it wasn't the worst either, I still enjoyed it :)

I have no idea how credit card numbers are generated. However, I've worked on a tangentially-related toy problem, if you're interested: how to go pseudo-randomly through all the possible numbers exactly once, without having to keep a list of previously-allocated numbers. And it uses a C++ standard class: [std::linear_congruential_engine](https://en.cppreference.com/w/cpp/numeric/random/linear_congruential_engine)! Blog post: [A Detour into Group Theory](https://vzalzal.com/posts/detour-group-theory/).

**Reply by elyashiv** [reply]
In a crypto course I have taken they suggested doing just that (using a generator in a cyclic group) for credit cards numbers.
There is an added constrain that you need the card to have a valid checksum. The solution was to advance the generator until you get a valid number.

**Comment by thomasgt** [comment]
[Patrick McKenzie](https://x.com/patio11) would be an interesting guest to discuss the intersection of finance and computing. He was recently on Conversations with Tyler [here](https://conversationswithtyler.com/episodes/patrick-mckenzie/)

**Comment by kentf** [comment]
Conor, sadly, I have to agree with Bryce, this episode was gold 😄.

As they say about streaming services/cable tv, "come for the movies, stay for the shows" your podcast is "come for the algorithms, stay for the banter" 😀.

"A discussion with Sean Parent"-episodes are always platinum-grade content!

Thanks!

**Comment by pethil** [comment]
This was fun. The high pitched forwarding noise really really hurts my ears. Please stop it. 🫰

**Reply by pethil** [reply]
Also I would have liked to just listen to my Brian rambling. Brian was right.

**Comment by Prof-RS** [comment]
Bryce, it will definitely be one of my favorite episodes of the year! 
Conner, listenable content wise, it was awful. However, the ~90 seconds of 10x audio was outstandingly hilarious. You even 10x’ed yourself!

**Comment by abeham** [comment]
Conor says there's a billion numbers per one of the 100,000 issuers which should last for 800 years and Bryce argues "People make a lot of babies". From a comedy standpoint that is gold. And I assume none of that was scripted. Overexcited Bryce and underwhelmed Conor is often very funny, but this episode took it to a whole new level. Even the editing of the 10x parts was gold. I mean, do you really just 10x and if yes, could the orginal be recovered by playing it at 1/10x speed? Not that I'd want to do that, but technically?

To answer your question "should we be buying ANSI standards or ISO standards once a month?"
It's your podcast, do whatever you want. I tune in for the other stuff that you usually chat about and discussing C++ standard proposals was a tiny bit more intereting for me, if you both can't let off the idea.

As for the title, technically it should probably read "... Credit Card Numbers", but that won't save this episode either.

So, thanks for the laughs, made my day!

**Comment by ninkibah** [comment]
Conor and Bryce,
when I was in Rwanda, we wrote a bare-bones acquirer system. I think what we did was allow the issuer to set up card products. So, you might have a standard Visa card, a platinum visa card, etc...
Then each product would be given a tranche of numbers to use for new cards. The bank could decide on how many numbers were in a tranche, and what the start and end numbers are.

I also remember taking the bits for those numbers and inverting them. This makes it difficult to predict the next number, as in an 8 bit mask, the numbers 1, 2 and 3 become 128, 64 and 192.

Hope that helps, and happy new year to the two of you.

**Comment by sputnick1124** [comment]
@codereport, implementing Luhn feels like a Rust vs C++ vs BQN vs Uiua video just waiting to happen. Here's a  [Uiua](https://uiua.org/pad?src=0_8_0__THVobiDihpAgPTDil78xMC8r4pmt4o2cKMKw4oqf4o2JKeKKmSgrw5fCrzk-OS7DlzIp4qyaMOKGr8KvMV8y4oeMCgotLS0KUGFyc2Ug4oaQIOKNoyjiiLXii5Up4ouF4ouFW13iipTilr3iiaBAIC4KRXhwZWN0ZWQg4oaQIFsxIDEgMSAxIDEgMSAxIDEgMCAwIDEgMF0KVGVzdHMg4oaQIOKItSjilqFQYXJzZSl7CiAgIjE3ODkzNzI5OTc0IgogICI0MzY5IDA0MjYgMjYxMyA1NTEwIgogICI0MzY5IDA0MzcgNjA4OCAwNzg4IgogICI0MzY5IDA0MjMgMDYwOCAzNDAyIgogICI0MzY5IDA0NTYgODEzNSA0MzE0IgogICI0NTM5IDMxOTUgMDM0MyA2NDY3IgogICIwNTkiCiAgIjA1NSA0NDQgMjg1IgogICIwNTUgNDQ0IDI4NiIKICAiODI3MyAxMjMyIDczNTIgMDU2OSIKICAiMDk1IDI0NSA4OCIKICAiMDU1IDQ1NCAyODUifQriiaEo4o2k4oqD4oiY4omNIOKKlEx1aG4pIFRlc3RzIEV4cGVjdGVkCi0tLQo=) implementation to get you started lol.
